### PR TITLE
feat: optimize selection implement by binary search

### PR DESCRIPTION
**Pull request by LucasXu0**

### CODE REVIEW

File: frontend/app_flowy/packages/flowy_editor/example/lib/main.dart [hunk 1]
```diff
@@ -205,4 +216,34 @@ class _MyHomePageState extends State<MyHomePage> {
       child: TextField(),
     );
   }
+
+  Widget _buildFlowyEditorWithBigDocument() {
+    return FutureBuilder<String>(
+      future: rootBundle.loadString('assets/big_document.json'),
+      builder: (context, snapshot) {
+        if (!snapshot.hasData) {
```
Comment (appflowy) [inline frontend/app_flowy/packages/flowy_editor/example/lib/main.dart:224]: I think we should write the normal path through the code first; then write the unusual cases.

if (snapshot.hasData) {
// normal path
} else {
}

File: frontend/app_flowy/packages/flowy_editor/lib/service/selection_service.dart [hunk 2]
```diff
@@ -359,12 +359,14 @@ class _FlowySelectionState extends State<FlowySelection>
...
           panStartOffsetWithScrollDyGap.translate(0, panStartScrollDy! - dy);
     }
-    final nodes = getNodesInRange(panStartOffsetWithScrollDyGap, panEndOffset!);
-    if (nodes.isEmpty) {
-      return;
-    }
-    final first = nodes.first.selectable;
-    final last = nodes.last.selectable;
+
+    final sortedNodes =
```
Comment (appflowy) [inline frontend/app_flowy/packages/flowy_editor/lib/service/selection_service.dart:363]: Nice try!

**Review by appflowy — APPROVED** [review]

File: frontend/app_flowy/packages/flowy_editor/example/lib/main.dart (hunk 1 above)
Comment (LucasXu0) [reply frontend/app_flowy/packages/flowy_editor/example/lib/main.dart:224]: I think we should put the code which's body is shorter before.

Comment (appflowy) [reply frontend/app_flowy/packages/flowy_editor/example/lib/main.dart:224]: Well, extracting the else condition into a function call would be better.

Comment (appflowy) [reply frontend/app_flowy/packages/flowy_editor/example/lib/main.dart:224]: Also, I found the e-version of the code complete. https://nikola-breznjak.com/blog/books/programming/code-complete-2-steve-mcconnell-using-conditionals/

Comment (LucasXu0) [reply frontend/app_flowy/packages/flowy_editor/example/lib/main.dart:224]: Got it.

**Pull request merged**